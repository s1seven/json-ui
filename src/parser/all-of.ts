import omit from "lodash/omit";
import mapValues from "lodash/mapValues";
import isArray from "lodash/isArray";
import isObject from "lodash/isObject";
import type { ExtractObjects, JSONSchema7Value } from "../utils/helper-types";
import { deepMerge } from "../utils/deep-merge";

export const allOf = <T extends JSONSchema7Value>(item: T): T =>
  !isObject(item)
    ? item
    : isArray(item)
    ? (item.map(allOf) as T)
    : (mapValues(
        isArray(item.allOf)
          ? deepMerge<ExtractObjects<JSONSchema7Value>>(
              ...(item.allOf.map(allOf) as ExtractObjects<JSONSchema7Value>[]),
              omit(item, "allOf")
            )
          : item,
        allOf
      ) as T);
