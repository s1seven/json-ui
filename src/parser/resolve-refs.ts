import {
  assign,
  get,
  isArray,
  isObject,
  isString,
  mapValues,
  omit,
} from "lodash";
import { JSONSchema7Value } from "../utils/helper-types";
import { REF_KEY } from "../constants";
import { JSONSchema7 } from "json-schema";

export const resolveRefs = (schema: JSONSchema7): JSONSchema7 => {
  const cache = (() => {
    const cacheMap = new Map<string, JSONSchema7>();
    return (path: string) => {
      if (cacheMap.has(path)) return cacheMap.get(path);
      const res = get(schema, path);
      cacheMap.set(path, res);
      return res;
    };
  })();

  const resolve = <T extends JSONSchema7Value>(item: T): T =>
    !isObject(item)
      ? item
      : isArray(item)
      ? (item.map(resolve) as T)
      : (mapValues(
          assign(
            omit(item, REF_KEY),
            isString(item[REF_KEY]) ? cache(refToPath(item[REF_KEY])) : {}
          ),
          resolve
        ) as T);
  return resolve(schema);
};

export const refToPath = (ref: string) =>
  ref.replace(/^#\/|\/+/g, ".").replace(/^\./, "");
