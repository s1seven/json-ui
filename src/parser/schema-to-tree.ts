import { JSONSchema7 } from "json-schema";
import { inferType } from "./infer";

export type Tree = (string | Tree)[];
// export type Tree = string[] | [string, Tree];
// export type Tree = (string | Tree)[];

// const tree: Tree = ["school", ["person", ["name", "age"]], "car"];

export const schemaToTree = (schema: JSONSchema7): Tree => {
  if (typeof schema === "string") {
    return [schema];
  } else if (Array.isArray(schema)) {
    const resultArray: Tree = [];
    for (const item of schema) {
      resultArray.push(schemaToTree(item));
    }
    return resultArray;
  } else if (typeof schema === "object") {
    const resultArray: Tree = [];
    for (const key in schema.properties) {
      if (schema.properties[key]) {
        resultArray.push(key);
        resultArray.push(schemaToTree(schema.properties[key] as JSONSchema7));
      }
    }
    return resultArray;
  }
  return [];
};

export const schemaToTree2 = (schema: JSONSchema7): Tree | undefined => {
  const type = inferType(schema);
  return type === "object"
    ? Object.entries(schema.properties!).map(([k, value]) => {
        const children = schemaToTree2(value as JSONSchema7);
        return children ? [k, children] : k;
      })
    : void 0;
};
