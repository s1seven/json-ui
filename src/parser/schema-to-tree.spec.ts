import { schemaToTree2 } from "./schema-to-tree";
import coaSchema from "../../fixtures/coa-schema.json";
import { resolveRefs } from "./resolve-refs";
import { JSONSchema7 } from "json-schema";
import { allOf } from "./all-of";

test("schema to tree", async () => {
  const schema = allOf(resolveRefs(coaSchema as any as JSONSchema7));

  console.log("my schema", schema);

  const tree = schemaToTree2(schema);
  console.log("my tree", tree);
  expect(tree).toMatchSnapshot();
});

// export const asdasd: TreeItem = [
//   "Root",
//   true,
//   [
//     ["Root.1", true],
//     ["Root.2", true, [["Root.2.1", true]]],
//     ["Root.1", true],
//   ],
// ];
