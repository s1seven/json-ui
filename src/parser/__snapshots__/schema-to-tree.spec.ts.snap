// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schema to tree 1`] = `
[
  "RefSchemaUrl",
  [
    "Certificate",
    [
      "CertificateLanguages",
      "Id",
      "Date",
      [
        "Standard",
        [
          "Norm",
          "Type",
        ],
      ],
      "Contacts",
      [
        "Parties",
        [
          [
            "Manufacturer",
            [
              "ZipCode",
              "City",
              "Country",
              "Email",
              "AdditionalInformation",
              "Street",
              [
                "Identifiers",
                [
                  "CageCode",
                ],
              ],
            ],
          ],
          [
            "Customer",
            [
              "ZipCode",
              "City",
              "Country",
              "Email",
              "AdditionalInformation",
              "Street",
              [
                "Identifiers",
                [
                  "CageCode",
                ],
              ],
            ],
          ],
          [
            "Receiver",
            [
              "ZipCode",
              "City",
              "Country",
              "Email",
              "AdditionalInformation",
              "Street",
              [
                "Identifiers",
                [
                  "CageCode",
                ],
              ],
            ],
          ],
          [
            "GoodsReceiver",
            [
              "ZipCode",
              "City",
              "Country",
              "Email",
              "AdditionalInformation",
              "Street",
              [
                "Identifiers",
                [
                  "CageCode",
                ],
              ],
            ],
          ],
        ],
      ],
      [
        "BusinessTransaction",
        [
          [
            "Order",
            [
              "Id",
              "Position",
              "Date",
              "Quantity",
              "QuantityUnit",
              "CustomerProductId",
              "CustomerProductName",
              "GoodsReceiptId",
            ],
          ],
          [
            "OrderConfirmation",
            [
              "Id",
              "Date",
            ],
          ],
          [
            "Delivery",
            [
              "Id",
              "Position",
              "Date",
              "Quantity",
              "QuantityUnit",
              "InternalOrderId",
              "InternalOrderPosition",
              "Transport",
            ],
          ],
        ],
      ],
      [
        "Product",
        [
          "Id",
          "Name",
          "CountryOfOrigin",
          "PlaceOfOrigin",
          "FillingBatchId",
          "FillingBatchDate",
          "ProductionBatchId",
          "ProductionDate",
          "Standards",
          "ExpirationDate",
          "AdditionalInformation",
        ],
      ],
      [
        "Analysis",
        [
          "LotId",
          "PropertiesStandard",
          "Inspections",
          "AdditionalInformation",
        ],
      ],
      [
        "DeclarationOfConformity",
        [
          "Declaration",
          [
            "CE",
            [
              "CE_Image",
              "NotifiedBodyNumber",
              "YearDocumentIssued",
              "DocumentNumber",
            ],
          ],
        ],
      ],
      "Attachments",
      "Logo",
      "Disclaimer",
    ],
  ],
]
`;
